platform/linux-generic/pktio/mdev.c

```diff
@@ -0,0 +1,585 @@
+
+int mdev_dma_area_alloc(mdev_device_t *mdev, mdev_dma_area_t *dma_area)
+{
+	struct vfio_iommu_type1_dma_map req;
+	void *tmp;
+
+	/* Make sure we're page aligned */
+	if (dma_area->size != ROUNDUP_ALIGN(dma_area->size, ODP_PAGE_SIZE))
+		return -EINVAL;
+
+	memset(&req, 0, sizeof(req));
+	req.argsz = sizeof(req);
+
+	tmp = mmap(NULL, dma_area->size, PROT_READ | PROT_WRITE,
+		   MAP_SHARED | MAP_ANONYMOUS | MAP_NORESERVE | MAP_FIXED, -1,
+		   0);
+	if (tmp == MAP_FAILED) {
+		ODP_ERR("mmap failed\n");
+		return -EFAULT;
+	}
+
+	dma_area->vaddr = (uint64_t)tmp;
+
+	req.vaddr = dma_area->vaddr;
+	req.size = dma_area->size;
+	req.flags = VFIO_DMA_MAP_FLAG_READ | VFIO_DMA_MAP_FLAG_WRITE;
+
+	if (ioctl(mdev->device, VFIO_IOMMU_MAP_DMA, &req) < 0) {
+		ODP_ERR("ioctl failed\n");
+		return -EFAULT;
+	}
+
+	dma_area->iova = req.iova;
+
+	ODP_DBG("dma_area alloc: %llx@%llx -> %llx\n", dma_area->size,
+		dma_area->vaddr, dma_area->iova);
+
+	return 0;
+}
+
+int mdev_dma_area_free(mdev_device_t *mdev, mdev_dma_area_t *dma_area)
+{
+	struct vfio_iommu_type1_dma_unmap req;
+
+	memset(&req, 0, sizeof(req));
+	req.argsz = sizeof(req);
+	req.iova = dma_area->iova;
+	req.size = dma_area->size;
+	req.flags = VFIO_DMA_MAP_FLAG_READ | VFIO_DMA_MAP_FLAG_WRITE;
+
+	if (ioctl(mdev->device, VFIO_IOMMU_UNMAP_DMA, &req) < 0) {
+		ODP_ERR("ioctl failed\n");
+		return -EFAULT;
+	}
+
+	if (munmap((void *)dma_area->vaddr, dma_area->size) < 0) {
+		ODP_ERR("munmap failed\n");
+		return -EFAULT;
+	}
+
+	ODP_DBG("dma_area free: %llx@%llx -> %llx\n", dma_area->size,
+		dma_area->vaddr, dma_area->iova);
+
+	return 0;
+}
+
+static char *mdev_basename(char *path)
+{
+	char *rpath;
+
+	rpath = basename(path);
+	if (rpath)
+		return rpath;
+
+	return NULL;
+}
+
+static int mdev_readlink(const char *path, char *link, size_t linksz)
+{
+	ssize_t len;
+
+	len = readlink(path, link, linksz - 1);
+	if (len != -1) {
+		link[len] = '\0';
+		return 0;
+	}
+	return -1;
+}
+
+/**
+ * returns group_id or -1 on fail and fills group_uuid
+ */
+static int mdev_sysfs_discover(const char *mod_name, const char *if_name,
+			       char *uuid, size_t sz)
+{
+	int ret;
+	char *driver, *iommu_group;
+	char sysfs_path[2048], sysfs_link[2048];
+	DIR *dir;
+	struct dirent *dp;
+
+	/* Don't put / on the end of the path */
+	snprintf(sysfs_path, sizeof(sysfs_path),
+		 "/sys/class/net/%s/device/driver", if_name);
+	ret = mdev_readlink(sysfs_path, sysfs_link, sizeof(sysfs_link));
+	if (ret) {
+		ODP_ERR("Can't locate sysfs driver path\n");
+		return -1;
+	}
+
+	driver = mdev_basename(sysfs_link);
+	if (!driver) {
+		ODP_ERR("Can't driver in sysfs\n");
+		return -1;
+	}
+
+	if (strcmp(driver, mod_name)) {
+		ODP_ERR("Invalid driver name\n");
+		return -1;
+	}
+
+	snprintf(sysfs_path, sizeof(sysfs_path),
+		 "/sys/class/net/%s/device/mdev_supported_types/%s-netmdev/devices/",
+		 if_name, driver);
+
+	dir = opendir(sysfs_path);
+	if (!dir)
+		return -1;
+	/* FIXME only the last uuid will be returned now */
```

Comment:
What is needed to resolve/add this? Is this intended to be a follow-on PR?

Reply:
This will need a new PR for multiple card support. we need to associate the NICs name(eth0/esp4s0 etc), with the appropriate uuid that was created. For now we assume only one card will be plugged in as a mediated device so we just return the last entry of the directory(which is actually the only entry)

Reply:
Actually, there shall be exactly one UUID per kernel netdevice as we have it like that on the kernel part right now. I'm in doubt that it will change ... So maybe we can just drop the FIXME.

Reply:
Either would be good. This can always be enhanced in follow-on PRs, but each PR should stand alone.

Reply:
Thanks, updated code according to an assumption that we have just one file in the directory which is UUID we're looking for.